p4-16/psa/PSA.mdk

```diff
@@ -2392,3 +2412,138 @@ out of range includes:
+to distinguish them.
+
+
+## No output port change during egress {#appendix-rationale-egress-cannot-change-output-port}
+
+Question: Why can't my P4 program change the output port during egress
+processing?
+
+In a network device that has many input and output ports, packets can
+arrive at or near the same time on multiple input ports, all destined
+for the same output port.
+
+Packet buffers are typically designed into such network devices, to
+store the packets that cannot be sent out immediately, absorbing this
+short term congestion.
+
+For a given output port P, we now wish to retrieve packets from the
+packet buffer at a rate that is equal to the rate we will send them to
+port P, typically equal to the maximum bit rate that it is possible to
+send data out of port P.
+
+Packet scheduling algorithms such as weighted fair queueing, and many
+others, have been developed that can determine which among a set of
+potentially many FIFO queues that a packet should be read from next,
+and sent out on the port.
+
+These link scheduling algorithms are real time algorithms with very
+tight timing constraints.  If they go too slow, the output port goes
+idle and its capacity is wasted.  If they go too fast, we read packets
+from the packet buffer faster than they can be transmitted on the
+port, and we are back at the same problem we had originally -- either
+drop some of the packets, or store them somewhere again until the port
+is ready to transmit them.
+
+Such a scheduling algorithm that handles multiple output ports must
+know which output port all packets are destined to, before they are
+put into the packet buffer.  If that target output port can be changed
+after the packet is read out, then we can simultaneously overload one
+output port while starving another.
+
+That is why the `egress_port` of a packet must be selected during
+ingress processing, and egress processing is not allowed to change it.
+
+These scheduling algorithms also need to know the size of each packet,
+i.e. the size as it will be when transmitted on the port.
+
+It is possible in egress P4 code to drop a packet, or to change the
+size of the packet by adding or removing headers.  Very likely
+P4-programmable network devices will have their scheduling algorithms
+run just slightly faster than the port to handle cases where many
+packets in a row are decreased in size during egress processing, and
+have tight control loops monitoring the size of packets leaving egress
+processing to make small adustments in the rate that the scheduling
+algorithm operates for each port.  Either that, or they will just
+leave some fraction of the output port's capacity unused during times
+when all packet sizes are being decreased.
+
+Certainly if there are long durations of time when egress decides to
+drop all packets to an output port, that port will go idle.  The
+scheduling algorithm implementations are all built with a finite
+maximum packet scheduling rate.
```

Comment:
Illuminating (for me). Thanks for putting this together. We might end up expanding it.

Reply:
No problem.  It is an issue that has come up a couple of times, and nice to have a place to give an answer in a little more than a phrase, but a lot less than 10 pages.